File: mobile/lib/features/products/domain/entities/product.dart
import 'package:equatable/equatable.dart';

class Product extends Equatable {
  const Product({
    required this.id,
    required this.sku,
    required this.name,
    required this.unit,
    required this.salePrice,
    this.categoryId,
    this.imagePath,
    this.onHand = 0,
    this.currencyCode,
    this.currencySymbol,
    this.boxQty,
    this.boxWeight,
    this.boxDimensions,
    this.saleMode = 'piece',
  });

  final int id;
  final String sku;
  final String name;
  final String unit;
  final double salePrice; // retail price (agents never see the purchase price)
  final int? categoryId;
  final String? imagePath;
  final double onHand; // available stock

  // Money type + packaging + sale mode. Set by a manager on web; agents only
  // read these here.
  final String? currencyCode;   // UZS, USD
  final String? currencySymbol; // so'm, $
  final int? boxQty;            // units per box; null = no box
  final double? boxWeight;      // kg per box
  final String? boxDimensions;  // "40x30x25 cm"
  final String saleMode;        // box | piece | both

  bool get hasBox => boxQty != null && boxQty! > 0;

  @override
  List<Object?> get props => [
        id,
        sku,
        name,
        unit,
        salePrice,
        categoryId,
        imagePath,
        onHand,
        currencyCode,
        currencySymbol,
        boxQty,
        boxWeight,
        boxDimensions,
        saleMode,
      ];
}

File: backend/app/schemas/product.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.enums import SaleMode


class ProductBase(BaseModel):
    # use_enum_values so sale_mode round-trips as its string value ("piece"),
    # matching the plain-text column, both when dumping to the model and to JSON.
    model_config = ConfigDict(use_enum_values=True)

    sku: str
    name: str
    barcode: str | None = None
    description: str | None = None
    unit: str = "шт"
    category_id: int | None = None
    cost_price: Decimal = Decimal("0")
    sale_price: Decimal = Decimal("0")
    min_stock: Decimal = Decimal("0")

    # Money type + packaging + sale mode. Manager-editable; agents read-only.
    currency_id: int | None = None
    box_qty: int | None = None
    box_weight: Decimal | None = None
    box_dimensions: str | None = None
    sale_mode: SaleMode = SaleMode.PIECE
    integer_qty: bool = True  # whole units only vs fractional (kg/litres)


class ProductCreate(ProductBase):
    pass


class ProductUpdate(BaseModel):
    model_config = ConfigDict(use_enum_values=True)

    sku: str | None = None
    name: str | None = None
    barcode: str | None = None
    description: str | None = None
    unit: str | None = None
    category_id: int | None = None
    cost_price: Decimal | None = None
    sale_price: Decimal | None = None
    min_stock: Decimal | None = None
    is_active: bool | None = None
    currency_id: int | None = None
    box_qty: int | None = None
    box_weight: Decimal | None = None
    box_dimensions: str | None = None
    sale_mode: SaleMode | None = None
    integer_qty: bool | None = None


class ProductOut(ProductBase):
    model_config = ConfigDict(from_attributes=True, use_enum_values=True)

    id: int
    is_active: bool
    image_path: str | None = None
    image_back_path: str | None = None
    on_hand: Decimal | None = None  # filled by the endpoint when stock is joined
    # Purchase price — hidden (None) for agents, who only see the retail price.
    cost_price: Decimal | None = None
    # Convenience currency fields, filled by the endpoint from the relationship.
    currency_code: str | None = None
    currency_symbol: str | None = None


class ProductHistoryEntry(BaseModel):
    kind: str  # "added" (stock-in by an account) | "sale" (sold by an agent)
    date: datetime
    quantity: Decimal
    user_name: str | None = None  # account that added stock, or agent that sold
    detail: str | None = None  # movement type, or "#<order> · <status>"
